gustaf/ffd.py

```diff
@@ -1,4 +1,627 @@
-"""gustaf/gustaf/ffd.py
+"""
+gustaf/gustaf/ffd.py
 
 Freeform Deformation!
+
+
+Adaptation of previous implementation in internal python package gustav by
+Jaewook Lee. 
 """
+from typing import List, Optional, Union
+import numpy as np
+from gustaf._base import GustafBase
+from gustaf.faces import Faces
+from gustaf.spline.base import Bezier, RationalBezier
+from gustaf.show import show_vedo
+from gustaf._typing import SPLINE_TYPES, MESH_TYPES
+from gustaf.create.spline import with_bounds
+
+
+class FFD (GustafBase):
+
+    def __init__(
+        self,
+        mesh: MESH_TYPES,
+        spline: Optional[SPLINE_TYPES],
+        # is_partial: bool=False,
+    ):
+        """
+        FreeFormDeformation. Here, spline is always the boss. Meaning, once
+        spline is defined using setter, we always fit mesh into the spline.
+        For partial FFD, no one needs to be the boss. If no spline is defined,
+        a simple spline wraps given mesh as default.
+
+        Parameters
+        ----------
+        mesh: MESH_TYPES
+            Base mesh of the FFD
+        spline: SPLINE_TYPE
+            Original Spline of the FFD
+        is_partial: bool
+            True if mesh is only partially deformed (currently not supported)
+
+        Attributes
+        ----------
+        _spline: SPLINE_TYPES
+        _mesh: MESH_TYPES
+        _original_spline_ranges: List[List[float]]
+        _q_vertices: np.ndarray (n, dim)
+        _is_partial: bool
+        calculated: bool
+
+        Returns
+        -------
+        None
+        """
+        # Use property definitions to store the values
+        self._spline = None
+        self._mesh: MESH_TYPES = None
+        self._original_spline_ranges = None
+        self._q_vertices = None
+        self._is_partial = False
+        self.calculated = False
+
+        self.mesh = mesh
+        if spline is None:
+            # if no spline is given create a very naive spline with physical
+            # bounds defined bny the mesh and parametric bounds defined by a 
+            # hypercube
+            self.spline = with_bounds(
+                [[0]*self._q_vertices.shape[1],[1]*self._q_vertices.shape[1]],
+                 self._mesh.bounds)
+        else:
+            self.spline = spline
+        self.is_partial = False
+
+
+
+        self._check_dimensions()
+
+    
+    def _check_dimensions(self):
```

Comment:
Yes mainly since there was at some point a discussion to check if the spline geometry dimension has the same bounds as the mesh. But I was unsure if this is even really possible to check for deformed splines bounds since these should in some cases be outside the bounds of the original mesh since it should deform the mesh into the new bounds.